operations/kubernetes/Containerd.md

```diff
@@ -1,99 +1,164 @@
+## `containerd` slow startup after reboot
 
+On some systems, `containerd` can take a very long time to start after a reboot. This has been fixed in CSM 1.3, but if this symptom occurs,
+messages indicating `cleaning up dead shim` may appear in the `containerd` log files. For example:
 
-### Restart containerd
+```text
+Aug 26 00:06:10 ncn-w001 containerd[4005]: time="2022-08-26T00:06:10.522985910Z" level=info msg="cleaning up dead shim"
+Aug 26 00:06:10 ncn-w001 containerd[4005]: time="2022-08-26T00:06:10.556198245Z" level=warning msg="cleanup warnings time=\"2022-08-26T00:06:10Z\" level=info msg=\"starting signal loop\" namespace=k8s.io pid=57627\n"
+Aug 26 00:06:10 ncn-w001 containerd[4005]: time="2022-08-26T00:06:10.556821890Z" level=info msg="loading plugin \"io.containerd.monitor.v1.cgroups\"..." type=io.containerd.monitor.v1
+Aug 26 00:06:10 ncn-w001 containerd[4005]: time="2022-08-26T00:06:10.557576058Z" level=info msg="loading plugin \"io.containerd.service.v1.tasks-service\"..." type=io.containerd.service.v1
+```
 
-If the containerd service is restarted on a worker node, this may cause the sonar-jobs-watcher pod running on that worker node to fail when attempting to cleanup unneeded containers. For example:
+Instructing `containerd` to remove shims when `containerd` is being shutdown will correct this issue.
 
-1. Restart containerd.
+1. Edit the `/srv/cray/resources/common/containerd/containerd.service` file.
+
+    Add the following `ExecStopPost` line to the file:
+    
+    ```text
+    ExecStopPost=/usr/bin/find /run/containerd/io.containerd.runtime.v2.task -name address -type f -delete
+    ```
+
+    After the edit, the relevant section of the file should look similar to the following:
+    ```text
+    [Service]
+    ExecStartPre=/sbin/modprobe overlay && /sbin/modprobe br_netfilter
+    ExecStart=/usr/local/bin/containerd
+    ExecStopPost=/usr/bin/find /run/containerd/io.containerd.runtime.v2.task -name address -type f -delete
+    Restart=always
+    RestartSec=5
+    Delegate=yes
+    ```
+
+1. Restart `containerd` to pick up the change.
+
+    > If this is a worker NCN, then also see the notes in the [Restarting `containerd` on a worker NCN](#restarting-containerd-on-a-worker-ncn)
+    > section for subsequent steps that must be taken after `containerd` is restarted.
 
     ```bash
-    ncn-w001# systemctl restart containerd
+    ncn-mw# systemctl restart containerd
     ```
 
-1.  Retrieve the name of the sonar-jobs-watcher pod.
+**NOTE:** If this NCN is rebuilt, then this change will need to be re-applied (until the system is upgraded to CSM 1.3).
+
+## Restarting `containerd` on a worker NCN
+
+If the `containerd` service is restarted on a worker node, then this may cause the `sonar-jobs-watcher` pod running on that worker node to fail when attempting to cleanup unneeded containers. Use the following procedure to determine if this is the case and how to remediate it.
```

Comment:
```suggestion
If the `containerd` service is restarted on a worker node, then this may cause the `sonar-jobs-watcher` pod running on that worker node to fail when attempting
to cleanup unneeded containers. Use the following procedure to determine if this is the case and how to remediate it.
```